x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/TypeResolutions.java

```diff
@@ -118,6 +118,25 @@ public static TypeResolution isIPAndExact(Expression e, String operationName, Pa
         return isExact(e, operationName, paramOrd);
     }
 
+    public static TypeResolution isStringOrDate(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
```

Comment:
It's rarely a good idea to touch the QL classes.
In this case instead of adding a specialized method in QL (which is used just once), use the following one liner in AutoBucket relies on `TypeResolutions#isType()` method (see the declarations of isDate/isIP/isString/ etc...:
```
TypeResolutions.isType(e, exp -> { DataTypes.isString(exp) || DataTypes.isDateTime(exp) }, operationName, paramord, "datetime", "string');
```

Reply:
As a side-note, QL projects rely on a centralize pluggable utility to perform conversions yet this is not leverages through-out the ESQL codebase (see `EsqlDataTypeConverter`) - this needs to change but in separate PR for the future https://github.com/elastic/elasticsearch/issues/104596

Reply:
Moved isStringOrDate from TypeResolutions to AutoBucket.